lib/librte_regexdev/rte_regexdev.c

```diff
@@ -2,4 +2,133 @@
  * Copyright(C) 2019 Marvell International Ltd.
  */
 
-#include <rte_regexdev.h>
+#include <string.h>
+
+#include <rte_spinlock.h>
+#include <rte_memory.h>
+#include <rte_memcpy.h>
+#include <rte_memzone.h>
+#include <rte_string_fns.h>
+
+#include "rte_regexdev.h"
+#include "rte_regexdev_driver.h"
+
+/* Shared memory between primary and secondary processes. */
+static struct {
+	uint64_t next_owner_id;
+	rte_spinlock_t ownership_lock;
+	struct rte_regex_dev_data data[RTE_MAX_REGEXDEV_PORTS];
+} *rte_regex_dev_shared_data;
+
+static const char *MZ_RTE_REGEX_DEV_DATA = "rte_regex_dev_data";
+
+struct rte_regex_dev rte_regex_devices[RTE_MAX_REGEXDEV_PORTS];
+
+/* spinlock for shared data allocation */
+static rte_spinlock_t rte_regex_shared_data_lock = RTE_SPINLOCK_INITIALIZER;
+
+
+static uint16_t
+regex_dev_find_free_port(void)
+{
+	unsigned i;
+
+	for (i = 0; i < RTE_MAX_REGEXDEV_PORTS; i++) {
+		if (rte_regex_devices[i].data != NULL &&
+		    rte_regex_devices[i].data->name == '\0')
+			return i;
+	}
+	return RTE_MAX_REGEXDEV_PORTS;
+}
+
+static struct rte_regex_dev*
+regex_dev_allocated(const char *name)
+{
+	unsigned i;
+
+	for (i = 0; i < RTE_MAX_REGEXDEV_PORTS; i++) {
+		if (rte_regex_devices[i].data != NULL &&
+		    strcmp(rte_regex_devices[i].data->name, name) == 0)
+			return &rte_regex_devices[i];
+	}
+	return NULL;
+}
+
+static struct rte_regex_dev*
+regex_dev_get(uint16_t port_id)
+{
+	struct rte_regex_dev *regex_dev = &rte_regex_devices[port_id];
+
+	regex_dev->data = &rte_regex_dev_shared_data->data[port_id];
+
+	return regex_dev;
+}
+
+static void
+regex_dev_shared_data_prepare(void)
+{
+	const unsigned flags = 0;
+	const struct rte_memzone *mz;
+
+	rte_spinlock_lock(&rte_regex_shared_data_lock);
+
+	if (rte_regex_dev_shared_data == NULL) {
+		if (rte_eal_process_type() == RTE_PROC_PRIMARY) {
+			/* Allocate port data and ownership shared memory. */
+			mz = rte_memzone_reserve(MZ_RTE_REGEX_DEV_DATA,
+					sizeof(*rte_regex_dev_shared_data),
+					rte_socket_id(), flags);
+		} else
+			mz = rte_memzone_lookup(MZ_RTE_REGEX_DEV_DATA);
+		if (mz == NULL)
+			rte_panic("Cannot allocate regex shared data\n");
+
+		rte_regex_dev_shared_data = mz->addr;
+		if (rte_eal_process_type() == RTE_PROC_PRIMARY) {
+			rte_spinlock_init(&rte_regex_dev_shared_data->ownership_lock);
+			memset(rte_regex_dev_shared_data->data, 0,
+			       sizeof(rte_regex_dev_shared_data->data));
+		}
+	}
+
+	rte_spinlock_unlock(&rte_regex_shared_data_lock);
+}
+
+struct rte_regex_dev*
+rte_regex_dev_register(const char *name)
+{
+	uint16_t port_id;
+	struct rte_regex_dev *regex_dev = NULL;
```

Comment:
lets try to follow reverse christmas tree scheme, so that we have to remember less coding conventions. :-)

Reply:
this is not a coding convention in DPDK